File: packages/shared/src/types/devis.ts
import { BaseEntity, UUID } from './common';

export type DevisStatut = 'brouillon' | 'envoye' | 'accepte' | 'refuse' | 'expire';

export interface LigneDevis {
  id: UUID;
  description: string;
  quantite: number;
  unite: string;
  prixUnitaireHT: number;
  tva: number; // En pourcentage (ex: 10, 20)
  montantHT: number;
  montantTTC: number;
}

export interface Devis extends BaseEntity {
  chantierId: UUID;
  numero: string; // ex: DEV-2025-001
  // Dates
  dateEmission: Date;
  dateValidite: Date;
  // Contenu
  lignes: LigneDevis[];
  // Totaux
  totalHT: number;
  totalTVA: number;
  totalTTC: number;
  // Statut
  statut: DevisStatut;
  dateAcceptation?: Date;
  signatureClient?: string;
  // PDF
  pdfUrl?: string;
  // Notes
  conditionsParticulieres?: string;
  notes?: string;
}

export interface CreateDevisDto {
  chantierId: UUID;
  dateValidite: Date;
  lignes: Omit<LigneDevis, 'id' | 'montantHT' | 'montantTTC'>[];
  conditionsParticulieres?: string;
  notes?: string;
}

export interface UpdateDevisDto {
  dateValidite?: Date;
  lignes?: Omit<LigneDevis, 'id' | 'montantHT' | 'montantTTC'>[];
  statut?: DevisStatut;
  conditionsParticulieres?: string;
  notes?: string;
}

File: apps/mobile/lib/domain/models/devis.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import '../enums/devis_statut.dart';

part 'devis.freezed.dart';
part 'devis.g.dart';

@freezed
abstract class Devis with _$Devis {
  const factory Devis({
    required String id,
    required String chantierId,
    required String numero,
    required DateTime dateEmission,
    required DateTime dateValidite,
    required double totalHT,
    required double totalTVA,
    required double totalTTC,
    @Default(DevisStatut.brouillon) DevisStatut statut,
    DateTime? dateAcceptation,
    String? signatureClient,
    String? pdfUrl,
    String? conditionsParticulieres,
    String? notes,
    DateTime? deletedAt,
    required DateTime createdAt,
    required DateTime updatedAt,
  }) = _Devis;

  factory Devis.fromJson(Map<String, dynamic> json) => _$DevisFromJson(json);
}
